indexing-service/src/main/java/io/druid/indexing/worker/WorkerTaskMonitor.java

```diff
@@ -91,151 +96,328 @@ public WorkerTaskMonitor(
+        return;
+      }
+
+      log.info("Submitting runnable for task[%s]", task.getId());
+
+      workerCuratorCoordinator.updateAnnouncement(
+          TaskAnnouncement.create(
+              task,
+              TaskStatus.running(task.getId()),
+              TaskLocation.unknown()
+          )
+      );
+
+      log.info("Affirmative. Running task [%s]", task.getId());
+      workerCuratorCoordinator.unannounceTask(task.getId());
+      final ListenableFuture<TaskStatus> future = taskRunner.run(task);
+      addRunningTask(task, future);
+    }
+  }
+
+  private class StatusNotice implements Notice
+  {
+    private final Task task;
+    private final TaskStatus status;
+
+    public StatusNotice(Task task, TaskStatus status)
+    {
+      this.task = task;
+      this.status = status;
+    }
+
+    @Override
+    public String getTaskId()
+    {
+      return task.getId();
+    }
+
+    @Override
+    public void handle()
+    {
+      final TaskDetails details = running.get(task.getId());
+
+      if (details == null) {
+        log.warn("Got status notice for task [%s] that isn't running...", task.getId());
+        return;
+      }
+
+      if (!status.isComplete()) {
+        log.warn(
+            "Got status notice for task [%s] that isn't complete (status = [%s])...",
+            task.getId(),
+            status.getStatusCode()
+        );
+        return;
+      }
+
+      details.status = status.withDuration(System.currentTimeMillis() - details.startTime);
+
+      try {
+        workerCuratorCoordinator.updateAnnouncement(
+            TaskAnnouncement.create(
+                details.task,
+                details.status,
+                details.location
+            )
+        );
+        log.info(
+            "Job's finished. Completed [%s] with status [%s]",
+            task.getId(),
+            status.getStatusCode()
+        );
+      }
+      catch (Exception e) {
+        log.makeAlert(e, "Failed to update task announcement")
+           .addData("task", task.getId())
+           .emit();
+      }
+      finally {
+        running.remove(task.getId());
+      }
+    }
+  }
+
+  private class LocationNotice implements Notice
+  {
+    private final String taskId;
+    private final TaskLocation location;
+
+    public LocationNotice(String taskId, TaskLocation location)
+    {
+      this.taskId = taskId;
+      this.location = location;
+    }
+
+    @Override
+    public String getTaskId()
+    {
+      return taskId;
+    }
+
+    @Override
+    public void handle()
+    {
+      final TaskDetails details = running.get(taskId);
+
+      if (details == null) {
+        log.warn("Got location notice for task [%s] that isn't running...", taskId);
+        return;
+      }
+
+      details.location = location;
+
+      try {
+        log.info("Updating task [%s] announcement to with location [%s]", taskId, location);
+        workerCuratorCoordinator.updateAnnouncement(
```

Comment:
nit: you could skip this announcement if details.location was already same as location.